src/algebra/module_regular.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.iterate_hom
+import algebra.module.basic
+
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+
+namespace module
+
+variables {R S : Type*} {a b : R} {s : S} (M : Type*)
+
+section has_scalar
+
+variables [has_scalar R M] [has_scalar R S] [has_scalar S M] [is_scalar_tower R S M]
+
+/-- An `M`-regular element is an element `c` such that multiplication on the left by `c` is an
+injective map `M → M`. -/
+def is_regular (c : R) := function.injective ((•) c : M → M)
+
+section monoid
+
+--variables [monoid R]
+--semigroup does not have the right instance
+--instance semigroup.has_scalar : has_scalar R R := { smul := λ a b, a * b }
+
+/-- In a monoid, the product of `M`-regular elements is `M`-regular. -/
+lemma is_regular.mul (ra : is_regular M a) (rs : is_regular M s) :
+  is_regular M (a • s) :=
+λ a b ab, rs (ra ((smul_assoc _ _ _).symm.trans (ab.trans (smul_assoc _ _ _))))
+
+/--  If an element `b` becomes `M`-regular after multiplying it on the left by an `M`-regular
+element, then `b` is `M`-regular. -/
+lemma is_regular.of_mul (ab : is_regular M (a • s)) :
+  is_regular M s :=
+@function.injective.of_comp _ _ _ (λ m : M, a • m) _ (λ c d cd, ab
+  (by rwa [smul_assoc, smul_assoc]))
+
+/--  An element is `M`-regular if and only if multiplying it on the left by an `M`-regular element
+is `M`-regular. -/
+@[simp] lemma mul_is_regular_iff (b : S) (ha : is_regular M a) :
+  is_regular M (a • b) ↔ is_regular M b :=
+⟨λ ab, is_regular.of_mul M ab, λ ab, is_regular.mul M ha ab⟩
+
+variables [monoid R] [is_scalar_tower R R M]
+
+/--  Two elements `a` and `b` are `M`-regular if and only if both products `a • b` and `b • a`
+are `M`-regular. -/
+lemma is_regular_mul_and_mul_iff :
+  is_regular M (a • b) ∧ is_regular M (b • a) ↔ is_regular M a ∧ is_regular M b :=
```

Comment:
Using `•` when `a` and `b` have the same type is probably just confusing:
```suggestion
  is_regular M (a * b) ∧ is_regular M (b * a) ↔ is_regular M a ∧ is_regular M b :=
```
The proof won't change.

Reply:
If I change the \smuls to *, the proof no longer works and I get again I think some incompatibility between `has_scalar` and `mul_action`, although it is not entirely clear to me what goes wrong.

Reply:
Odd. I think the way to proceed is to add these lemmas too:
```
lemma is_regular.mul (ra : is_regular M a) (rb : is_regular M b) : is_regular M (a * b) :=
ra.smul rb

lemma is_regular.of_mul (ab : is_regular M (a * b)) :
  is_regular M b :=
by { rw ←smul_eq_mul at ab, exact ab.of_smul }

@[simp] lemma is_regular.mul_iff  (ha : is_regular M a) :
  is_regular M (a * b) ↔ is_regular M b :=
⟨is_regular.of_mul, ha.mul⟩
```

Reply:
Actually, with the new class, this seems to work as is.  I will wait for CI to build and see if it requires more changes!

Reply:
I still think this lemma should be stated in terms of `mul`

Reply:
Ah, I think that I changed a different one, convinced that it was this one!  Now I changed it and the proof did indeed work with no change!